modules/core/src/main/java/org/apache/ignite/internal/processors/query/QueryUtils.java

```diff
@@ -640,7 +640,7 @@ public static void processBinaryMeta(GridKernalContext ctx, QueryEntity qryEntit
             d.addProperty(prop, false);
         }
 
-        if (!isKeyClsSqlType)
+        if (!isKeyClsSqlType && qryEntity instanceof QueryEntityEx && ((QueryEntityEx)qryEntity).isPreserveKeysOrder())
```

Comment:
Which cases qryEntity couldn't be an instance of  QueryEntytyEx?

Reply:
In case a table created from Java API. Actually it COULD be, but it SHOULDN'T be. But for non-sql tables this is not a problem anyway since the unwrap exists only for tables created via SQL API

Reply:
As of now yes, in the future seems to unwrap could be extended to all cases. However, it will be an issue for another ticket.